### PR TITLE
Get Maps by Type

### PR DESCRIPTION
**Pull request by moxplod**
# Description

Added a method in mapper to return the maps by type. This can be used to define custom value convertors by the caller. 

Discussion - https://github.com/alirezanet/Gridify/discussions/213
Related issue - #211 

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

- [x] I have performed a self-review of my code
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have made corresponding changes to the documentation
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: src/Gridify/GridifyMapper.cs [hunk 1]
```diff
@@ -220,6 +220,26 @@ public IEnumerable<IGMap<T>> GetCurrentMaps()
...
+      foreach (var map in _mappings)
+      {
+         if (map.To.Body is UnaryExpression unaryExpression)
+         {
+            if (targetTypes.Contains(unaryExpression.Operand.Type))
+            {
+               yield return map;
+            }
+         }
+         // TODO: need to implement other types that aren't unary expression
```
Comment (alirezanet) [inline src/Gridify/GridifyMapper.cs:235]: This needs to be completed to support and test all types; currently, it only works with unary expressions. Adding a couple of additional unit tests for known dotnet types and a few custom types is also required.

**Review by alirezanet — CHANGES_REQUESTED** [review]

File: src/Gridify/GridifyMapper.cs (hunk 1 above)
Comment (moxplod) [reply src/Gridify/GridifyMapper.cs]: I ppdated the method a bit further, I am not sure what other types of expressions are out there. Where can I find the list? 

Probably my lack of depth in understanding expressions. This approach isn't looking so hot anymore, where we have to figure out each expression type to find the property type.

Comment (alirezanet) [reply src/Gridify/GridifyMapper.cs]: Handling all types of expressions isn't necessary; I believe these three cover at least 80% of the types, and so far, it looks very promising. Perhaps we should consider adding another test case to determine if we can support custom types. Lastly, the PR seems fine, but the documentation needs to be updated.

File: test/Gridify.Tests/GridifyMapperShould.cs [hunk 2]
```diff
@@ -149,8 +149,31 @@ public void AddMap_DuplicateKey_ShouldThrowErrorIfOverrideIfExistsIsFalse()
       //The thrown exception can be used for even more detailed assertions.
       Assert.Equal("Duplicate Key. the 'test' key already exists", exception.Message);
    }
+   
+   [Theory]
+   [InlineData(typeof(DateTime), 0)]
```
Comment (alirezanet) [inline test/Gridify.Tests/GridifyMapperShould.cs:154]: one improvement would be to add a test case that has a DateTime ... false test is not so useful

**Review by alirezanet — APPROVED** [review]
Looks good,
Just documentation, and if you have time more tests with other types

File: test/Gridify.Tests/GridifyMapperShould.cs (hunk 2 above)
Comment (alirezanet) [reply test/Gridify.Tests/GridifyMapperShould.cs:154]: one other thing that I just remembered is nested mappings... does it work with nested ones? plus, mappings that has a `Select` method for collections. we need to test more before merging it

File: src/Gridify/GridifyMapper.cs [hunk 3]
```diff
@@ -220,6 +220,39 @@ public IEnumerable<IGMap<T>> GetCurrentMaps()
    {
       return _mappings;
    }
+   
+   public IEnumerable<IGMap<T>> GetCurrentMapsByType(HashSet<Type> targetTypes)
+   {
+      foreach (var map in _mappings)
+      {
+         if (map.To.Body is UnaryExpression unaryExpression)
```
Comment (alirezanet) [inline src/Gridify/GridifyMapper.cs:228]: minor improvement, we can convert this to a switch statement (not important)

Comment (moxplod) [reply src/Gridify/GridifyMapper.cs:228]: Done

**Pull request merged**